### ISSUE TITLE
Security considerations for receiving videos

### ISSUE THREAD

**Issue by patrickab** [timokau/prefq#67]
At the moment anyone can easily send any kind of data to the server. This data will then be stored locally, and offers and easy entrance for malware. Considering some sort of possibility for authentification would be nice at this point. An easy method would be to implement an optional argument during the initialization of the server, that allows the user to specify a password. Then, the server could only accept POST requests to `/video`, that contain this password. A corresponding functionality could easily be added to the Query Client class. If provided with a password during object initialization, it simply adds this password to the POST request before sending it to the server.

**Comment by timokau** [comment]
Thank you for raising this issue! In general, all query client access should be authenticated. I am not sure if a password in the POST request is the best way of achieving this, maybe we should look into existing methods and their benefits and drawbacks (such as risk of password sniffing).

**Comment by patrickab** [comment]
Good point, as mentioned, its probably the easiest way, but still not very secure. I dug a bit into the topic, and JSON Web Token (JWT) is a promising option. Also, there is a flask extension, called [Flask-JWT](https://pypi.org/project/Flask-JWT/), which allows for easy implementation of secure communication.

```python
from flask import Flask, request
from flask_jwt import JWT, jwt_required, current_identity
from datetime import datetime, timedelta

app = Flask(__name__)
app.config['API-KEY'] = 'very-secret-key'

class User(object):
    def __init__(self, id):
        self.id = id

def authenticate():

    # Your authentication logic goes here
    # For example, we could authenticate based on client IP, API key, etc.
    # If authentication is successful, return a User object

    api_key = request.headers.get('X-API-Key')  # Assuming the API key is provided in a header
    if api_key == app.config['API-KEY']:
        return User(id=some-id)
    else:
        return None  # Authentication failed, return None

def identity(payload):
    user_id = payload['identity']
    # Retrieve the user based on the user_id from the payload
    # This function is called after successful authentication to get the user object
    return User(id=user_id)

jwt = JWT(app, authenticate, identity)

# Token expiration time
app.config['JWT_EXPIRATION_DELTA'] = timedelta(days=1)

@app.route('/videos', methods=['POST'])
@jwt_required()
def receive_videos():
    # [...]

if __name__ == '__main__':
    app.run(debug=True)
```

The API-Key could be altered, whenever we enter a route, where JWT is required. This new key can then be returned to the Query Client.

However, this would practically require the Query Client to know the initial key and also to be executed until termination.

**Comment by timokau** [comment]
JWT sounds promising. I think it is not a big limitation that it requires that the query client knows the key ahead of time, since query client and server will likely be set up by the same person or at least the responsible people will have a secure channel of communication.

Could you come up with a proposal PR?

**Comment by patrickab** [comment]
I dug a bit deeper into the topic, one problem that would still the need for encrypted message transfer, as otherwise the API-Key, which grants access to the Web Token, would also be vulnerable to password sniffing. Consequently, the requests should also be encrypted.

I decided to implement encryption via SSH keypairs to start solving this issue (see Pull Request #71)

Implementing this procedure allows us to encrypt the api-password as first layer of security. This procedure in theory allows password sniffing, as an attacker could extract the encrypted password from an intercepted request. However, if we manage to also encrypt the videos, then the server will also decrypt them with the private key, making a potential attack harmless. This means, that as long as the public key remains unknown, the proposed method is secure.

I have also made an attempt to use https, but unfortunately I still have problems with the OCI free trial. However, I gathered some knowledge and information about open-source generation for SSL Certificates using [Certbot](https://certbot.eff.org/), and a python library for SSL-Certificates compatible with both flask and waitress. This is a promising starting point for implementing https communication in future.
